### PR TITLE
`mutate()` preserves attributes of grouped data frames

### LINKED ISSUE #6100: data.frame attributes are preserved on `mutate()` but dropped on `group_by |> mutate`
Reprex:
``` r
library(dplyr)
#>
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#>
#>     filter, lag
#> The following objects are masked from 'package:base':
#>
#>     intersect, setdiff, setequal, union
library(purrr)
attr(mtcars, "test") <- "foo"
mtcars_grouped <- group_by(mtcars, cyl)

mtcars |>
  mutate(new_col = 1) |>
  attributes() |>
  pluck("test")
#> [1] "foo"

mtcars_grouped |>
  mutate(new_col = 1) |>
  attributes() |>
  pluck("test")
#> NULL

mtcars_grouped |>
  mutate(across(where(is.complex), as.character)) |>
  attributes() |>
  pluck("test")
#> NULL
# did nothing but still lost attrs on the base data.frame
```

It would be better if no attributes were lost with mutate. Feels weird in the last case where no mutate actually gets done.

However if `group_by() |> mutate()` must drop attributes, it's probably better that `mutate` does also. You can easily get tricked into thinking some code is going to work, but then it bombs when it accidentally gets passed some sticky groups. This happened to me today, 4 levels of package context up from where the mutate was.

### PR DESCRIPTION
**Pull request by romainfrancois**
closes #6100 

``` r
library(dplyr, warn.conflicts = FALSE)

library(purrr)
attr(mtcars, "test") <- "foo"
mtcars_grouped <- group_by(mtcars, cyl)

mtcars %>% 
  mutate(new_col = 1) %>% 
  attr("test")
#> [1] "foo"

mtcars_grouped %>% 
  mutate(new_col = 1) %>% 
  attr("test")
#> [1] "foo"
```

<sup>Created on 2021-11-29 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

### CODE REVIEW

File: tests/testthat/test-count-tally.r [hunk 1]
```diff
@@ -54,8 +54,7 @@ test_that("output preserves class & attributes where possible", {
   out <- df %>% group_by(g) %>% count()
   expect_s3_class(out, "grouped_df")
   expect_equal(group_vars(out), "g")
-  # summarise() currently drops attributes
```
Comment (romainfrancois) [inline tests/testthat/test-count-tally.r:57]: As the comment suggest, I believe the test was wrong.

**Comment by hadley** [comment]
Are you thinking of this for 1.0.8 or 1.1.0?

**Comment by romainfrancois** [comment]
I believe this should should be 1.0.8, and then we can think of something more intentional for 1.1.0: inform or warn about unknown attributes, hint about creating a custom class that deals with them, ... 

What we have right now "keeping attributes only for ungrouped data frames" is surprising, as in #6100

**Comment by DavisVaughan** [comment]
I do wonder if we should hold off on this until dplyr 1.1.0. We have the same issue with `dplyr_row_slice()`

``` r
library(dplyr)

df <- data.frame(x = 1)
gdf <- group_by(df, x)

attr(df, "keepme") <- "foo"
attr(gdf, "keepme") <- "foo"

# kept
attr(slice(df, 1), "keepme")
#> [1] "foo"

# dropped
attr(slice(gdf, 1), "keepme")
#> NULL
```

I think the whole generic API here needs another round of thinking through

**Comment by romainfrancois** [comment]
🤔 yeah perhaps as a follow up to this, `dplyr_row_slice.*` should all call `dplyr_reconstruct_core()` on the result of `vec_slice()` and then reconstruct the `groups`. 

But otherwise, I don't mind waiting for 1.1 so that we can design this more.

**Comment by hadley** [comment]
Yeah, I'd rather we wait, especially since we're close to release.

**Comment by hadley** [comment]
Let's hold off on this as it's looking like `.by` might provide a simpler alternative.

**Pull request closed without merging**